### ISSUE TITLE
Is it karmada support deploy in IPv4/IPv6 dual-stack mode

### ISSUE THREAD

**Issue by SZT0728** [karmada-io/karmada#2569]
is  karmada:v1.2.0 support deploy in  IPv4/IPv6 dual-stack mode?
if not, which version support?

**Comment by RainbowMango** [comment]
Could you please tell us more details about your case? I'm afraid I can't get your point.

**Comment by SZT0728** [comment]
> Could you please tell us more details about your case? I'm afraid I can't get your point.

i want to install karmada in the kubernetes which network-mo de,and join the ipv6 k8s cluster

**Comment by RainbowMango** [comment]
To be honest, I'm not sure since I never tested it before. The only difference between IPv4 and IPv6 is the address of kube-apiserver, given we are using client-go(i bet it supports IPv6) to talk to the member cluster, so I guess it should be OK.

Can you help to test it and give us feedback? (I don't have the environment on hand.)

**Comment by SZT0728** [comment]
> To be honest, I'm not sure since I never tested it before. The only difference between IPv4 and IPv6 is the address of kube-apiserver, given we are using client-go(i bet it supports IPv6) to talk to the member cluster, so I guess it should be OK.
> 
> Can you help to test it and give us feedback? (I don't have the environment on hand.)

the etcd start error with the log: unexpected error setting up listen-peer-urls: URL address does not have the form "host:port": http://fd74:ca9b:172:18:63cf:ab87:ceb8:1807:2380
how can i deal with it?

**Comment by RainbowMango** [comment]
I guess this error was raised during the installation, right? which tool are you using? Karmadactl or chart?
Could you please show more logs here?

**Comment by SZT0728** [comment]
> I guess this error was raised during the installation, right? which tool are you using? Karmadactl or chart? Could you please show more logs here?

yes, it was raise when i exec `kubectl karmada init --crds=crdspath`. i use the tool names  `kubectl-karmada`. no more logs,just this line.
I get the etcd statefulSet in yaml and the etcd exec command with args ` listen-peer-urls: http://${POD_IP}:2380`.  But the pod ip is ipv6-mode,it should identify like `http://[${POD_IP}]:2380`.  
Is this problem because ETCD does not handle ipv6 format？

**Comment by RainbowMango** [comment]
Do you have a Kubernetes running in IPv6 mode? Can you show me what's the `--listen-peer-urls` look like?

We just fixed a similar issue by #2450 last week. 
@duanmengkk Have you ever met this issue?

**Comment by SZT0728** [comment]
> Do you have a Kubernetes running in IPv6 mode? Can you show me what's the `--listen-peer-urls` look like?
> 
> We just fixed a similar issue by #2450 last week. @duanmengkk Have you ever met this issue?

The Kubernetes running in IPv6 mode show the etcd args `- --listen-peer-urls=https://[4001:250:250:250:250:250:250:222]:2380`
 I recompile a kubectl-karmada according to [https://github.com/karmada-io/karmada/pull/2450/commits/bd6b872a8d8965b983199ee72d3f686751412fc0](url)  to see if I can fix it

**Comment by RainbowMango** [comment]
Good luck, looking forward to your feedback.

**Comment by duanmengkk** [comment]
> Do you have a Kubernetes running in IPv6 mode? Can you show me what's the `--listen-peer-urls` look like?
> 
> We just fixed a similar issue by #2450 last week. @duanmengkk Have you ever met this issue?

No,I didn‘t met it.’May be we can open an issue such as `Support ipv6 in karmada` to track it ，rather than resolve it one by one

**Comment by SZT0728** [comment]
> Good luck, looking forward to your feedback.

I install the karmada in the IPv6 mode kubernetes and the etcd start failed with behind:
{"level":"fatal","ts":"2022-09-28T04:25:32.516Z","caller":"etcdmain/etcd.go:203","msg":"discovery failed","error":"--initial-cluster has etcd-0=http://etcd-0.etcd.karmada-system.svc.cluster.local:2380 but missing from --initial-advertise-peer-urls=http://[fd74:ca9b:172:18:ee5a:9c44:af76:2993]:2380 (\"http://[fd74:ca9b:172:18:ee5a:9c44:af76:2993]:2380\"(resolved from \"http://[**fd74:ca9b:172:18:ee5a:9c44:af76:2993**]:2380\") != \"http://**10.244.0.209**:2380\"(resolved from \"http://etcd-0.etcd.karmada-system.svc.cluster.local:2380\"))","stacktrace":"go.etcd.io/etcd/server/v3/etcdmain.startEtcdOrProxyV2\n\t/tmp/etcd-release-3.5.1/etcd/release/etcd/server/etcdmain/etcd.go:203\ngo.etcd.io/etcd/server/v3/etcdmain.Main\n\t/tmp/etcd-release-3.5.1/etcd/release/etcd/server/etcdmain/main.go:40\nmain.main\n\t/tmp/etcd-release-3.5.1/etcd/release/etcd/server/main.go:32\nruntime.main\n\t/home/remote/sbatsche/.gvm/gos/go1.16.3/src/runtime/proc.go:225"}

`fd74:ca9b:172:18:ee5a:9c44:af76:2993` and `10.244.0.209` are the same pod etcd-0's ip address in different mode.

and the etcd statefulSet's initContainers command like below:

`initContainers:
      - command:
        - sh
        - -c
        - "set -ex\ncat <<EOF | tee /etc/etcd/etcd.conf\nname: ${POD_NAME}\nclient-transport-security:\n
          \ client-cert-auth: true\n  trusted-ca-file: /etc/kubernetes/pki/ca.crt\n
          \ key-file: /etc/kubernetes/pki/etcd-server.key\n  cert-file: /etc/kubernetes/pki/etcd-server.crt\npeer-transport-security:\n
          \ client-cert-auth: false\n  trusted-ca-file: /etc/kubernetes/pki/ca.crt\n
          \ key-file: /etc/kubernetes/pki/etcd-server.key\n  cert-file: /etc/kubernetes/pki/etcd-server.crt\ninitial-cluster-state:
          new\ninitial-cluster-token: etcd-cluster\ninitial-cluster: etcd-0=http://etcd-0.etcd.karmada-system.svc.cluster.local:2380\nlisten-peer-urls:
          http://[${POD_IP}]:2380 \nlisten-client-urls: https://[${POD_IP}]:2379,http://127.0.0.1:2379\ninitial-advertise-peer-urls:
          http://[${POD_IP}]:2380\nadvertise-client-urls: https://${POD_NAME}.etcd.karmada-system.svc.cluster.local:2379\ndata-dir:
          /var/lib/etcd\n\n"`
how can i solve this problem?

**Comment by SZT0728** [comment]
> Good luck, looking forward to your feedback.

The karmada currently not support deploy in the dual-stack kubernetes well  because of the etcd. The etcd start failed with the follow:
`"error":"--initial-cluster has etcd-0=http://etcd-0.etcd.karmada-system.svc.cluster.local:2380 but missing from --initial-advertise-peer-urls=http://[fd74:ca9b:172:18:ee5a:9c44:af76:29aa]:2380 (\"http://[fd74:ca9b:172:18:ee5a:9c44:af76:29aa]:2380\"(resolved from \"http://[fd74:ca9b:172:18:ee5a:9c44:af76:29aa]:2380\") != \"http://10.244.0.232:2380\"(resolved from \"http://etcd-0.etcd.karmada-system.svc.cluster.local:2380\"))`

And this error cause by the etcd resolve domain return the first address default. [https://sourcegraph.com/github.com/etcd-io/etcd@v3.5.1/-/blob/pkg/netutil/netutil.go?L32:5](url)

So in the dual-stack kubernetes it return the IPv4 rather than the IPv6.

**Comment by RainbowMango** [comment]
Thanks for your analysis. Do you know what we can do here?
also cc author @prodanlabs for comment.

**Comment by duanmengkk** [comment]
The actual reason is the server of etcd is listened in ipv4 mode,see the etcd.conf ` kubectl logs -f etcd-0 -n karmada-system-test2 etcd-init-conf`
![image](https://user-images.githubusercontent.com/17697155/194750622-bf07cf57-3291-4134-bf3e-d62200c0b799.png)
However,the ip address resolved by core-dns is ipv6,  `nslookup etcd-0.etcd.karmada-system-test2.svc.cluster.local`
![image](https://user-images.githubusercontent.com/17697155/194750833-111ad2f0-49d2-4ac0-a9f1-181d1e92377d.png)

you can manually change the service of etcd to fix the problem temporarily.
```
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
```

**Comment by duanmengkk** [comment]
We can add a flag such as `karmadactl init --mode ipv6 `or dual-stack to fix it.But I'm not so sure that's a reasonable solution. @RainbowMango @prodanlabs 
The Helm Install  also have the same problem.https://github.com/karmada-io/karmada/blob/789a44acfaa71159a50e1c650b89a5746a15f1a8/charts/karmada/templates/etcd.yaml#L91

**Comment by SZT0728** [comment]
> The actual reason is the server of etcd is listened in ipv4 mode,see the etcd.conf ` kubectl logs -f etcd-0 -n karmada-system-test2 etcd-init-conf` ![image](https://user-images.githubusercontent.com/17697155/194750622-bf07cf57-3291-4134-bf3e-d62200c0b799.png) However,the ip address resolved by core-dns is ipv6, `nslookup etcd-0.etcd.karmada-system-test2.svc.cluster.local` ![image](https://user-images.githubusercontent.com/17697155/194750833-111ad2f0-49d2-4ac0-a9f1-181d1e92377d.png)
> 
> you can manually change the service of etcd to fix the problem temporarily.
> 
> ```
>   ipFamilies:
>   - IPv4
>   ipFamilyPolicy: SingleStack
> ```

sorry, Here's a bug I forgot to mention. I recompile the `kubectl-karmada` and add the symbol `[]` to the `listen-peer-urls`、`listen-client-urls`、`listen-client-advertise-peer-urls` for the IPv6 deploy. see the etcd config `kubectl logs -f etcd-0 -n karmada-system etcd-init-conf`
<img width="780" alt="截屏2022-10-11 下午2 53 54" src="https://user-images.githubusercontent.com/15276000/195017301-386d3982-1f47-454c-8d78-28496f87269b.png">

And the etcd service describe as below:
<img width="512" alt="截屏2022-10-11 下午2 57 16" src="https://user-images.githubusercontent.com/15276000/195017736-16d0b2d5-3dc2-4ada-805c-85d8913c9fcd.png">

and the etcd start failed due to the etcd resolve above, it resolve `http://etcd-0.etcd.karmada-system.svc.cluster.local:2380` return the first result default. And the first result is IPv4，not equal the `initial-advertise-peer-urls` which value in IPv6 mode

**Comment by SZT0728** [comment]
it translate the etcd-0 domain and the etcd resolve it and return the first result



发自我的iPhone


------------------ Original ------------------
From: duanmengkk ***@***.***&gt;
Date: Tue,Oct 11,2022 4:37 PM
To: karmada-io/karmada ***@***.***&gt;
Cc: SZT0728 ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [karmada-io/karmada] Is it karmada support deploy in IPv4/IPv6dual-stack mode (Issue #2569)





  
it resolve http://etcd-0.etcd.karmada-system.svc.cluster.local:2380 return the first result default
  
you mean nslookup etcd-0.etcd.karmada-system.svc.cluster.local returned a ipv4 address? In my env,it is a ipv6 address
